### PR TITLE
feat: debug_getRawTransaction RPC endpoint

### LINKED ISSUE #8152: Add getRawTransaction endpoint
### Component

Anvil

### Describe the feature you would like

Context (from the Telegram group):

> Guilherme Dantas, [13/06/2024 09:39]
> Hey, does anyone know if it is possible to retrieve blob content (blobs, KZG commitments and proofs) after sending blob-carrying txs to anvil?
> 
> Matthias Seitz, [13/06/2024 10:11]
> hmm, we're actually missing the various getRawTransaction endpoints
> could you please open an issue for this, these can be added easily
> 
> Matthias Seitz, [13/06/2024 10:13]
> that's the only way to retrieve them and no guarantee that the blobs still exist, but anvil also keeps them around after they've been mined so the getRawTransaction format should include them

### Additional context

_No response_

### PR DESCRIPTION
**Pull request by dorlneylon**
## Motivation

Closes #8152

## Solution

I'm not exactly sure about @mattsse reply:

> this includes endpoints:
> * `debug_getRawTransaction`
> * `eth_getRawTransactionByHash`
> * `eth_getRawTransactionByBlockHashAndIndex`
> * `eth_getRawTransactionByBlockNumberAndIndex`

Since, as shown here: https://ethereum.org/en/developers/docs/apis/json-rpc/ and here: https://geth.ethereum.org/docs/interacting-with-geth/rpc/ns-eth, non-eth endpoints should not have `eth_` prefix. Maybe move them to `debug_`?

### CODE REVIEW

File: crates/anvil/src/eth/api.rs [hunk 1]
```diff
@@ -1436,6 +1448,58 @@ impl EthApi {
...
+    ///
+    /// Handler for RPC call: `debug_getRawTransaction`
+    pub async fn raw_transaction(&self, hash: B256) -> Result<Option<Bytes>> {
+        node_info!("debug_getRawTransaction");
+
+        let mut out = vec![];
+
+        match self.pool.get_transaction(hash) {
+            Some(tx) => {
+                tx.transaction.encode_2718(&mut out);
```
Comment (mattsse) [inline crates/anvil/src/eth/api.rs:1461]:
```suggestion
                return Ok(Some(tx.transaction.encoded_2718()));
```

File: crates/anvil/src/eth/api.rs [hunk 2]
```diff
@@ -1436,6 +1448,58 @@ impl EthApi {
...
+            }
+            None => return Ok(None),
+        }
+
+        if !out.is_empty() {
+            Ok(Some(out.into()))
+        } else {
+            Ok(None)
+        }
```
Comment (mattsse) [inline crates/anvil/src/eth/api.rs:1470]: here we also need to check if the tx is mined,

similar to transaction_by_hash

**Review by mattsse — CHANGES_REQUESTED** [review]
tysm for this.

a few nits,

it'd be great to get a few simple tests for this.

1. retrieve pooled tx
2. retrieve mined tx

see this test for reference
https://github.com/foundry-rs/foundry/blob/41a6945ca3ef8139ee15c2c52986b085792be7fa/crates/anvil/tests/it/transaction.rs#L703-L703

File: crates/anvil/src/eth/api.rs (hunk 2 above)
Comment (dorlneylon) [reply crates/anvil/src/eth/api.rs:1470]: wdym "check if tx is mined"? the backend struct returns unsigned txs, and they can't be encoded till they appear in pool.

Comment (dorlneylon) [reply crates/anvil/src/eth/api.rs:1470]: though I can do this prolly:

```rust
match self.backend.transaction_by_hash(hash).await? {
    Some(tx) => {
        Ok(Some(self.sign_transaction(tx.into_transaction_request()).await?.into()))
    }
    None => Ok(None),
}
```

**Comment by dorlneylon** [comment]
@mattsse I have provided the tests, resolved the nits and refactored the code. Please recheck.

File: crates/anvil/src/eth/api.rs [hunk 3]
```diff
@@ -78,6 +78,25 @@ use futures::channel::{mpsc::Receiver, oneshot};
...
+        if $request.gas.is_none() {
+            // estimate if not provided
+            if let Ok(gas) = $self.estimate_gas($request.clone(), None, None).await {
+                $request.gas = Some(gas.to());
+            }
+        }
+
+        let $typed_tx = $self.build_typed_tx_request($request, $nonce)?;
+    };
+}
```
Comment (mattsse) [inline crates/anvil/src/eth/api.rs:98]: please remove this macro, this is very hard to read and unclear why this is needed.

you should be able to use

https://github.com/alloy-rs/alloy/blob/143f175d934b83619b02ae6dd820f5e145c64381/crates/rpc-types-eth/src/transaction/mod.rs#L259-L259

instead

**Review by mattsse — CHANGES_REQUESTED** [review]

File: crates/anvil/src/eth/api.rs (hunk 3 above)
Comment (dorlneylon) [reply crates/anvil/src/eth/api.rs]: well, actually it removes a lot of unnecessary code duplication in `EthApi`. Maybe I should think about its realisation more and make another PR...

But for now, I have removed it

**Comment by dorlneylon** [comment]
@mattsse fixed. Also please consider thinking on using the macro, since it reduces code duplication. Maybe I need to come up with another PR with the macro?..

**Review by mattsse — APPROVED** [review]

**Pull request merged**